File: Best_time_to_buy_and_sell_stocks(121).java
class Solution{
  public int maxProfit(int[] prices){
    int len=prices.length;
    int profit=0;
    int min=Integer.MAX_VALUE;
    for(int i=0; i<len; i++){
      if(profit[i]<min){
        min=profit[i];
      }
      profit=Math.max(profit, profit[i]-min);
    }
     return profit;
  }
}


// here, adding the "profit=Math.max(profit, profit[i]-min);" under the else block optimizes the solution

/*The else block ensures that profit calculation only happens when prices[i] is greater than or equal to minPrice.
  But since profit calculation (prices[i] - minPrice) already results in 0 when prices[i] == minPrice, the program works fine without else too.*/

File: Best_time_to_buy_and_sell_stocks(121).cpp
//C++ solution to the leetcode question 121

//Using else block which reduces runtime


class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int len=prices.size();
        int minimum=INT_MAX;
        int profit=0;
        for(int i=0; i<len; i++){
            if(prices[i]<minimum){
                minimum=prices[i];
            }
            else{ 
            profit=max(profit, prices[i]-minimum);
            }
            
        }
        return profit;

    }
};
